scripts/zephyr_module.py

```diff
@@ -182,37 +182,47 @@ def main():
                              list`""")
     parser.add_argument('-x', '--extra-modules', nargs='+', default=[],
                         help='List of extra modules to parse')
-    parser.add_argument('-w', '--west-path', default='west',
-                        help='Path to west executable')
     parser.add_argument('-z', '--zephyr-base',
                         help='Path to zephyr repository')
     args = parser.parse_args()
 
+
+    if args.zephyr_base is not None:
+        os.chdir(args.zephyr_base)
+
     if args.modules is None:
-        p = subprocess.Popen([args.west_path, 'list', '--format={posixpath}'],
-                             cwd=args.zephyr_base,
-                             stdout=subprocess.PIPE,
-                             stderr=subprocess.PIPE)
-        out, err = p.communicate()
-        if p.returncode == 0:
-            projects = out.decode(sys.getdefaultencoding()).splitlines()
-        elif re.match((r'Error: .* is not in a west installation\.'
-                       '|FATAL ERROR: no west installation found from .*'
-                       '|FATAL ERROR: no west workspace.*'),
-                      err.decode(sys.getdefaultencoding())):
-            # Only accept the error the event we are outside a west
-            # workspace.
-            #
-            # TODO: we can just use "west topdir" instead if we can
-            # depend on west 0.7.0 or later.
-            projects = []
-        else:
-            print(err.decode(sys.getdefaultencoding()))
-            # A real error occurred, raise an exception
-            raise subprocess.CalledProcessError(cmd=p.args,
-                                                returncode=p.returncode)
-    else:
-        projects = args.modules
+        # West is imported here, as it is optional (and thus maybe not installed)
+        # if user is providing a specific modules list.
+        from west.app.main import main as westmain
```

Comment:
Ugh, hold on, I am sorry about this, but, actually, looking at all this code here with fresh eyes, I have just realized that zephyr_module.py has no need to be running west as a subprocess at all. West has an API, and we can just use it directly. That was why you, others, and I went to all the trouble of implementing the program that way...

```
>>> from west.manifest import Manifest
>>> m = Manifest.from_file()
>>> project_paths = [p.posixpath for p in m.get_projects([])]
>>> project_paths
['/home/mbolivar/zp/zephyr', '/home/mbolivar/zp/modules/hal/cmsis', ...,
'/home/mbolivar/zp/tools/edtt', '/home/mbolivar/zp/modules/tee/tfm']
```

This is cleaner in my opinion and will be faster on Windows. Plus any west exceptions can be handled very specifically instead of catching SystemExit and redirecting west's standard err *and* out to stderr, mixing the streams.

You can see other ways to locate the manifest file when calling `Manifest.from_file()` in the docstring if you need to start from somewhere other than the current working directory.

I am using a funny `get_projects` invocation on purpose because of https://github.com/zephyrproject-rtos/west/issues/327, which we would want to be future-proof against. The important thing to remember is that the manifest repository might be a module (as it is with `nrf`).

When west 327 is implemented `m.projects` will not include the manifest repository, and so using a more readable `[p.posixpath for p in m.projects]` would silently drop the manifest repository module from the list. In our case, nrf is such an example where the manifest repository is a module, so we definitely don't want that to break ;). We can keep this future-proof by keeping `get_projects` around  but adding a keyword argument `include_manifest_repo` that defaults to `True` when we implement west 327, then deprecating it in favor of a method with another name slowly and over time.

Reply:
great, this is **much** nicer.
Will update code.

Reply:
done.
and now I actually like the code myself :smile: